OCaml_Intro_to_CS/part_6.ml

```diff
@@ -0,0 +1,220 @@
+(*
+Suppose you have two functions f : ‘a -> ‘b and g :
+‘c -> ‘a. Define their functional composition as follows:
+  *)
+
+let compose f g = fun x -> f (g x);;
+
+let inc x = x + 1;;
+
+let double x = x + x;;
+
+let h = compose inc double;;
+
+let result = h 5 ;;
+
+let () =
+  Printf.printf "result of compose inc double 5: %d\n" result;; 
+
+(*
+Please implement the compose_list function. Note: if the list is empty,
+then please return the argument (that is, be an identity function). In other words:
+# compose_list [] 7;;
+- : int = 7
+*)
+
+let dec x = x - 1;;
+
+let compose_list lst x =
+  let rec aux acc = function
+    | [] -> if acc = 0  then 
+      (
+        print_endline "Reached Identity";
+        x (* identity *)
+      )
+      else (
+        print_endline "Reached end of list without identity";
+        acc 
+      )
+    | h :: t -> 
+        if acc = 0 then 
+          (
+            print_endline "Applying function from list";
+            aux (h x) t 
+          )
+        else 
+          (
+            print_endline "Applying function from list with acc";
+            aux (h acc) t
+          )
+  in
+  aux 0 lst
+;;
+
+let operations = [inc ; double ; dec];;
+
+let res = compose_list operations 7;;
+
+let () = 
+  Printf.printf "Result of compose_list operations: %d\n" res;;
+
+(* for the extra 1 mark, using fold_left and fold_right *)
+let compose_list_fl func_lst x =
+  let f acc func = func acc in
+  let init = x in
+  List.fold_left f init func_lst
+;;
+
+let fl_res = compose_list_fl operations 7;;
+
+let () = 
+  Printf.printf "Result of compose_list_fl operations: %d\n" fl_res;;
+
+let compose_list_fr func_lst x =
+  let f func acc = func acc in
+  let init = x in
+  List.fold_right f func_lst init
+;;
+
+let fr_res = compose_list_fr operations 7;;
+
+let () = 
+  Printf.printf "Result of compose_list_fr operations: %d\n" fr_res;;
+
+(*
+Your task: implement mapi : (int -> 'a -> 'b) -> 'a list -> 'b list.
+  *)
+
+let mapi f lst =
+  let mapper i acc x =
+    let mapped_value = f i x in
+    mapped_value :: acc
+  in
+  let indexed_list = List.mapi (fun i x -> (i, x)) lst in
+  let init = [] in 
+  let mapped_values = List.fold_left (fun acc (i, x) -> mapper i acc x) init indexed_list in
+  List.rev mapped_values
+;;
```

Comment:
The new implementation is slightly better, you're doing 2 passes through the list instead of 3. The reason you're getting the reverse order is because you use an accumulator to generate to append the elements. To get the order in a single pass (at the expense of more memory on the stack) don't use tail recursion and do this

```ocaml
let mapi f lst = 
  let rec aux index = function
  | [] -> []
  | h :: t -> f index h :: aux (index + 1) t
  in
aux lst
```